### PR TITLE
UFT hyperparameter tuning 

### PR DESCRIPTION
**Pull request by wiinnie9**
- hyperparameter tuning using optuna library
- output best_params in yml and use it to perform unpaired training

### CODE REVIEW

File: train/uft/config.yml [hunk 1]
```diff
@@ -1,13 +1,5 @@
...
-step_lr_step_size: 30
-step_lr_gamma: 0.1
-batch_size: 4
-print_loss_interval: 100
-calc_eval_loss_interval: 100
-denoiser_loss_b1: 0.5
-denoiser_loss_b2: 0.5
-checkpoint_path: C:\Users\super\Documents\1.School\FYP\fyp_project\src\toenet\checkpoint\checkpoint.pth.tar
-images_dir: C:\Users\super\Documents\1.School\FYP\fyp_project\Data\input\unpaired
-save_dir: C:\Users\super\Documents\1.School\FYP\fyp_project\Data\output
```
Comment (superleesa) [inline train/uft/config.yml]: why are these parameters removed?

File: train/uft/tuning.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        dataloader,
+        save_dir,
+        checkpoint_path,
+        save_images= "sample"
+    )
+
+    avg_psnr = psnr_per_sample.mean().item()
+    avg_ssim = ssim_per_sample.mean().item()
+    
+    return avg_psnr + avg_ssim
```
Comment (superleesa) [inline train/uft/tuning.py]: actually we can't simply add these values because PSNR usally takes a reange of 30~50 while SSIM is always within a range 0 ~ 1. 

two possible solutions
1. normalize the PSNR value within the range of 0 to 1, for fair comparison
2. simply use PSNR only (using SSIM only might cause overfitting because we use SSIM as loss)

i guess we can do the second solution because it's simpler.

File: train/uft/tuning.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+def objective (trial):
+    config_path = Path(__file__).parent / "config.yml"
+    images_dir =  load_directory_from_yml(config_path)["images_dir"]
+    num_epochs = trial.suggest_int('num_epochs', 1, 2)
+    batch_size = trial.suggest_int('batch_size', 4, 24, step = 4)
+    train_ratio = 0.7
+    checkpoint_path = load_directory_from_yml(config_path)["checkpoint_path"]
+
+    save_dir = load_directory_from_yml(config_path)["save_dir"]
+    save_dir = create_unique_save_dir(save_dir)
```
Comment (superleesa) [inline train/uft/tuning.py]: can you chain these function calls?

File: train/uft/tuning.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            'denoiser_adversarial_loss_clip_min': trial.suggest_float('denoiser_adversarial_loss_clip_min', 0.0, 1.0),
+            'denoiser_adversarial_loss_clip_max': trial.suggest_float('denoiser_adversarial_loss_clip_max', 0.0, 1.0),
+            'calc_eval_loss_interval': trial.suggest_int("calc_eval_loss_interval", 50, 500),
+    }
+
+    return params
+
+def objective (trial):
+    config_path = Path(__file__).parent / "config.yml"
+    images_dir =  load_directory_from_yml(config_path)["images_dir"]
```
Comment (superleesa) [inline train/uft/tuning.py]: i think you are calling load_directory_from_yml multiple times in this function. can you call once and store all the variables at the same time? (you are accessing the disk everytime to load config)

**Review by superleesa — CHANGES_REQUESTED** [review]

File: train/uft/config.yml (hunk 1 above)
Comment (wiinnie9) [reply train/uft/config.yml]: all parameters are initialised in the tuning.py in the get_params(trial) function to be called in the optuna objective function

File: train/uft/tuning.py (hunk 2 above)
Comment (wiinnie9) [reply train/uft/tuning.py]: ok i change to PSNR only

File: train/uft/tuning.py (hunk 3 above)
Comment (wiinnie9) [reply train/uft/tuning.py]: ok can thanks

File: train/uft/tuning.py (hunk 4 above)
Comment (wiinnie9) [reply train/uft/tuning.py]: ok can

File: train/uft/config.yml (hunk 1 above)
Comment (wiinnie9) [reply train/uft/config.yml]: image

Comment (wiinnie9) [reply train/uft/config.yml]: the directories only would be in the config.yml

Comment (superleesa) [reply train/uft/config.yml]: i see but the original train.py for UFT uses those configurations and so deleting them cause the train.py to no longer work. can you can create a separate config file for the tuning pls?

Comment (superleesa) [reply train/uft/config.yml]: @weeenieee

**Review by superleesa — APPROVED** [review]
looks fine to me overall, but can you just change the part i mentioned in the above comments pls?

File: train/uft/config.yml (hunk 1 above)
Comment (wiinnie9) [reply train/uft/config.yml]: Ok sure, can

**Comment by wiinnie9** [comment]
Done, @superleesa can check again and merge. There's merging conflicts, I believe.
